### DISCUSSION TITLE
Calcolo dell'energia sballato con Shelly EM

### DISCUSSION THREAD

**Discussion by diegoves** [indomus/forum#2544] [HUB: Home Assistant]
Ciao a tutti! Dopo un po' di tribolamenti sono finalmente riuscito ad avere una prima plancia con i dati che mi servono per il monitoraggio dei consumi. HA OS l'ho installato sul QNAP tramite Virtualization Station seguendo la guida del blog, ed i vari sensori sono tutti Shelly EM. 
Ho configurato i vari sensori ed anche alcuni helper per avere il monitoraggio giornaliero, ed ora l'unico che mi sta dando problemi è quello relativo al consumo di energia generata dai pannelli. 
Il mio obiettivo è riuscire ad avere un monitoraggio di quanta energia consumo proveniente dai pannelli, quindi ho creato questo sensore:

      assorbita_pannelli_kwh:
        friendly_name: "Corrente assorbita dai pannelli (kWh)"
        unit_of_measurement: "kWh"
        value_template: >-
          {% if states('sensor.fotovoltaico_power')|float|round(3, default=0) > 0 %}
            {% if states('sensor.vano_scala_channel_1_power')|float|round(3, default=0) > 0 %} 
              {{ states('sensor.vano_scala_channel_2_energy_returned')|float|round(3, default=0) }}
            {% else %} 
              {{ states('sensor.fotovoltaico_energy')|float|round(3, default=0) }}
            {% endif %}
          {% else %} 
            {{ 0|float|round(0, default=0) }}
          {% endif %}
        icon_template: "mdi:flash"

dove:
-) sensor.fotovoltaico_power indica la potenza (W) generata in quell'istante dai pannelli;
-) sensor.vano_scala_channel_1_power indica la potenza (W) misurata a monte del contatore generale, se è positiva allora è corrente verso la rete, se è negativa allora è corrente verso casa;
-) vano_scala_channel_2_energy_returned indica l'energia (kWh), sempre positiva, consumata da tutto l'appartamento;
-) sensor.fotovoltaico_energy indica l'energia (kWh), sempre positiva, generata dal fotovoltaico. 

Il problema è che quando creo un utility meter su questo sensore, ci sono momenti (non ho ancora capito quando di preciso, ma sembra intorno a quando i pannelli cominciano a generare e quando smettono) in cui salta su a mille ed oltre kWh. Per dire, ora alle 10.47, pur avendo azzerato il contatore a mezzanotte, mi segna 231549kWh assorbiti dai pannelli. 
Cosa sto sbagliando?
(Tutti gli altri utility meter collegati direttamente ai sensori Shelly senza "calcoli" in mezzo, invece, funzionano perfettamente)

Grazie mille in anticipo!

**Comment by KC-inDomus** [comment]
segui queste best pratice: https://indomus.it/progetti/integrare-e-contabilizzare-un-pannello-solare-fotovoltaico-plug-and-play-su-home-assistant/

**Comment by Death0rz** [comment]
Così ad occhio direi che stai mischiando potenza (W) con energia (kWh). Premetto che sono abbastanza neofita ed ignorante in merito, ma sembra che appunto usi l'energia che è un contatore per contare la potenza che è invece un'entità a stati senza incremento.
Se ti può essere utile questi sono i miei sensori che ho creato per arrivare ad una situazione simile alla tua.
`template:
  - sensor:
      - name: "Immissione Rete"
        unique_id: immissione_rete
        state_class: measurement
        device_class: power
        unit_of_measurement: W
        state: >
          {%if states('sensor.potenza_fv')|float - states('sensor.casa_marco_channel_1_power')|float > 0 %}
              {{states('sensor.potenza_fv')|float - states('sensor.casa_marco_channel_1_power')|float}}
          {% else %}
              0
          {% endif %}
  - sensor:
      - name: "Prelievo Rete"
        unique_id: prelievo_rete
        state_class: measurement
        device_class: power
        unit_of_measurement: W
        state: >
          {%if states('sensor.potenza_fv')|float - states('sensor.casa_marco_channel_1_power')|float <= 0 %}
            {{(states('sensor.potenza_fv')|float - states('sensor.casa_marco_channel_1_power')|float)*-1}}
          {% else %}
            0
          {% endif %}`

Per avere il sensore di energia derivante dalla potenza di immissione in rete ho creato un helper con somma di Riemann della potenza immessa in rete. Così da poterla aggiungere sulla plancia Energia di HA
